### PR TITLE
Unit Test framework.go

### PR DESCRIPTION
**Pull request by cmicat**
Signed-off-by: raymondmiaochaoyue <raymondmiaochaoyue@didiglobal.com>

**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Add Unit Test for framework.go

**Which issue(s) this PR fixes**:

Part of https://github.com/karmada-io/karmada/issues/2579

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by cmicat** [comment]
<https://github.com/cmicat/karmada-io-karmada/actions/runs/3241704021/jobs/5314114312>

"Upload coverage to Codecov" Step fails consitently. Is there any problem in the test pipeline configuration?

```
[2022-10-13T11:03:46.018Z] ['info'] Pinging Codecov: https://codecov.io/upload/v4?package=github-action-3.1.1-uploader-0.3.2&token=*******&branch=test%2Fframework&build=3241704021&build_url=https%3A%2F%2Fgithub.com%2Fcmicat%2Fkarmada-io-karmada%2Factions%2Fruns%2F3241704021&commit=ba161e13be6731db39d0927bad22877c79a975fb&job=CI+Workflow&pr=&service=github-actions&slug=cmicat%2Fkarmada-io-karmada&name=&tag=&flags=unittests&parent=
[2022-10-13T11:03:46.018Z] ['verbose'] Passed token was 0 characters long
[2022-10-13T11:03:46.018Z] ['verbose'] https://codecov.io/upload/v4?package=github-action-3.1.1-uploader-0.3.2&branch=test%2Fframework&build=3241704021&build_url=https%3A%2F%2Fgithub.com%2Fcmicat%2Fkarmada-io-karmada%2Factions%2Fruns%2F3241704021&commit=ba161e13be6731db39d0927bad22877c79a975fb&job=CI+Workflow&pr=&service=github-actions&slug=cmicat%2Fkarmada-io-karmada&name=&tag=&flags=unittests&parent=
        Content-Type: 'text/plain'
        Content-Encoding: 'gzip'
        X-Reduced-Redundancy: 'false'
[2022-10-13T11:03:46.414Z] ['error'] There was an error running the uploader: Error uploading to [https://codecov.io:](https://codecov.io/) Error: There was an error fetching the storage URL during POST: 404 - {'detail': ErrorDetail(string='Could not find a repository, try using repo upload token', code='not_found')}
[2022-10-13T11:03:46.414Z] ['verbose'] The error stack is: Error: Error uploading to [https://codecov.io:](https://codecov.io/) Error: There was an error fetching the storage URL during POST: 404 - {'detail': ErrorDetail(string='Could not find a repository, try using repo upload token', code='not_found')}
    at main (/snapshot/repo/dist/src/index.js)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
[2022-10-13T11:03:46.414Z] ['verbose'] End of uploader: 554 milliseconds
Error: Codecov: Failed to properly upload: The process '/home/runner/work/_actions/codecov/codecov-action/v3/dist/codecov' failed with exit code 255
```

**Comment by RainbowMango** [comment]
I don't know, I also see the error sometimes.

**Comment by liangyuanpeng** [comment]
You need to set CODECOV_TOKEN  for this action if you want to run action on your fork repo.

**Comment by liangyuanpeng** [comment]
You can sett what how can i configuration for it: https://github.com/liangyuanpeng/karmada/commit/32f6ea29c45bf92f981325bf186c643a0998db2a#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR86

And doc from codecov-action: 

https://github.com/codecov/codecov-action

image

**Comment by cmicat** [comment]
> You need to set `CODECOV_TOKEN` for this action if you want to run action on your fork repo.

Actually, I don't need to run any action on my own fork repo. I only need to run this action automatically in merge request to get a "Codecov Report", like this: <https://github.com/karmada-io/karmada/pull/2631#issuecomment-1275726911>

**Comment by liangyuanpeng** [comment]
> Actually, I don't need to run any action on my own fork repo. I only need to run this action automatically in merge request to get a "Codecov Report", like this: https://github.com/karmada-io/karmada/pull/2631#issuecomment-1275726911
Absolutely.

>  https://github.com/cmicat/karmada-io-karmada/actions/runs/3241704021/jobs/5314114312
> "Upload coverage to Codecov" Step fails consitently. Is there any problem in the test pipeline configuration?
The reason I say this is because I saw that the link you sent was your fork repo.

**Comment by RainbowMango** [comment]
> For some unknown reason, github cannot generate "Codecov Report" for this Pull Request.

I think the service of [Codecov](https://app.codecov.io/) is unviable sometimes.

> Actually, I don't need to run any action on my own fork repo.

+1
We can ignore the upload thing from the fork repos.
Maybe we can skip it as follows:
```diff
$ git diff .github/workflows/ci.yml
diff --git a/.github/workflows/ci.yml b/.github/workflows/ci.yml
index 78d467c2..b3ebe2e9 100644
--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -81,6 +81,8 @@ jobs:
       - name: make test
         run: make test
       - name: Upload coverage to Codecov
+        # prevent running from forked repository
+        if: ${{ github.repository == 'karmada-io/karmada' }}
         uses: codecov/codecov-action@v3
``` 

@cmicat @liangyuanpeng any of you want to try it and send a PR if it works?

**Comment by liangyuanpeng** [comment]
> We can ignore the upload thing from the fork repos.
> Maybe we can skip it as follows:

Ha,  Actually, I have really work for it and pr is ready.  https://github.com/liangyuanpeng/karmada/pull/5/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR85

And then given that the contributors do want to run CI in their repositories, eg:  The first PR from Contributor. So i didn't continue to work on it. I'm trying to figure out if there's a better way.

**Comment by cmicat** [comment]
> > We can ignore the upload thing from the fork repos.
> > Maybe we can skip it as follows:
> 
> Ha, Actually, I have really work for it and pr is ready. https://github.com/liangyuanpeng/karmada/pull/5/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR85
> 
> And then given that the contributors do want to run CI in their repositories, eg: The first PR from Contributor. So i didn't continue to work on it. I'm trying to figure out if there's a better way.

Since I do not need to run action on my repo, I just disabled all actions in Github settings of my repository. But this will need every cloned repository owner to do this themselves.

**Comment by liangyuanpeng** [comment]
> I just disabled all actions in Github settings of my repository.

Some confused, Action of github fork repo is disable by default. BTW, create a issue for this topic would be great and this PR focus  on your contribution. :)

File: pkg/scheduler/framework/runtime/framework_test.go [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+	"github.com/karmada-io/karmada/pkg/scheduler/framework"
+)
+
+const (
+	justFilterPluginName = "justFilterPlugin"
+	justScorePluginName  = "justScorePlugin"
+	bothPluginName       = "bothPlugin"
+)
+
+type justFilterPlugin struct {
```
[the remark this answers is not in the document]
Comment (cmicat) [reply pkg/scheduler/framework/runtime/framework_test.go:23]: OK

File: pkg/scheduler/framework/runtime/framework_test.go [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+
+func (j *justFilterPlugin) Name() string {
+	return justFilterPluginName
+}
+
+func (j *justFilterPlugin) Filter(ctx context.Context, placement *policyv1alpha1.Placement, bindingSpec *workv1alpha2.ResourceBindingSpec, clusterv1alpha1 *clusterv1alpha1.Cluster) *framework.Result {
+	return framework.NewResult(j.code)
+}
+
+type justScorePlugin struct {
```
[the remark this answers is not in the document]
Comment (cmicat) [reply pkg/scheduler/framework/runtime/framework_test.go:41]: OK

File: pkg/scheduler/framework/runtime/framework_test.go [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+
+func (j *justScorePlugin) ScoreExtensions() framework.ScoreExtensions {
+	return j
+}
+
+func (j *justScorePlugin) NormalizeScore(ctx context.Context, scores framework.ClusterScoreList) *framework.Result {
+	return framework.NewResult(j.normalizeScoreCode)
+}
+
+type bothPlugin struct{}
```
[the remark this answers is not in the document]
Comment (cmicat) [reply pkg/scheduler/framework/runtime/framework_test.go:80]: OK

File: pkg/scheduler/framework/runtime/framework_test.go [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+			if got.IsSuccess() != tt.wantSuccess {
+				t.Errorf("RunFilterPlugins() = %v, want %v", got, tt.wantSuccess)
+			}
+		})
+	}
+}
+
+func Test_frameworkImpl_RunScorePlugins(t *testing.T) {
+	cluster1 := newCluster("cluster1")
+	cluster2 := newCluster("cluster2")
```
[the remark this answers is not in the document]
Comment (cmicat) [reply pkg/scheduler/framework/runtime/framework_test.go:204]: OK

File: pkg/scheduler/framework/runtime/framework_test.go [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+	fr, err := NewFramework(r)
+	if err != nil {
+		t.Errorf("NewFramework() returned error: %v", err)
+		return
+	}
+
+	impl := fr.(*frameworkImpl)
+
+	expectFilterPlugin(t, impl)
+	expectScorePlugin(t, impl)
```
[the remark this answers is not in the document]
Comment (cmicat) [reply pkg/scheduler/framework/runtime/framework_test.go:122]: I think you mean use a String Set? 

Like `k8s.io/apimachinery/pkg/util/sets/string.go` ?

[the remark this answers is not in the document]
Comment (RainbowMango) [reply pkg/scheduler/framework/runtime/framework_test.go:122]: yes. Sorry, I missed that part from the last post.

**Comment by RainbowMango** [comment]
Hi, for the coverage upload failure, I found this issue
https://github.com/codecov/codecov-action/issues/598

**Comment by RainbowMango** [comment]
I'm looking at #2654 which uses `gomock` to simulate the test. @cmicat would you like to take a look?

**Comment by cmicat** [comment]
> I'm looking at #2654 which uses `gomock` to simulate the test. @cmicat would you like to take a look?

Actually, I looked at <https://github.com/golang/mock> before writing this unit test. I think hand written test object is good enough for this test.

**Comment by RainbowMango** [comment]
> I think using an additional program to generate mocks, then write unit test is a little complex.

I agree with you, #2654 seems less straightforward enough compared with this patch.
But with gomock, we can use less code to simulate the interface behaviors.

Personally, I tend to the gomock approach. How do you think?

**Comment by cmicat** [comment]
> I agree with you, #2654 seems less straightforward enough compared with this patch. But with gomock, we can use less code to simulate the interface behaviors.
> 
> Personally, I tend to the gomock approach. How do you think?

I haven't used gomock myself. So I cannot comment is it easier to write or not.

In my experience with mockito, the code will be:

(1) shorter

(2) easier to read

(3) More complex to write in some conditions.

Because the behavior of mockito may be a little unexpected.

As for is this PR or <https://github.com/karmada-io/karmada/pull/2654> better. I think <https://github.com/karmada-io/karmada/pull/2654> is more clean to read. aka better.

**Comment by cmicat** [comment]
Since https://github.com/karmada-io/karmada/pull/2654 is merged, this PR can be closed.

**Comment by RainbowMango** [comment]
Thanks @cmicat all the same~

**Pull request closed without merging**